astropy/nddata/nddata.py

```diff
@@ -78,97 +69,122 @@ class NDData(NDDataBase):
     """
 
     def __init__(self, data, uncertainty=None, mask=None, wcs=None,
-                 meta=None, unit=None):
+                 meta=None, unit=None, copy=False):
 
+        # Rather pointless since the NDDataBase does not implement any setting
+        # but before this PR (#4234) the NDDataBase did call the uncertainty
+        # setter and if anyone wants to alter this behaviour again this call
+        # to the superclass NDDataBase should be in here.
         super(NDData, self).__init__()
 
+        # Check if data is any type from which to collect some implicitly
+        # passed parameters.
         if isinstance(data, NDData):  # don't use self.__class__ (issue #4137)
-            # No need to check the data because data must have successfully
-            # initialized.
-            self._data = data._data
-            self._unit = data.unit  # must set before uncert for NDDataArray
-            self.uncertainty = data.uncertainty
-            self._mask = data.mask
-            self._wcs = data.wcs
-            self._meta = data.meta
-
-            if uncertainty is not None:
-                self._uncertainty = uncertainty
-                log.info("Overwriting NDData's current uncertainty being"
-                         " overwritten with specified uncertainty")
-
-            if mask is not None:
-                self._mask = mask
+            # Of course we need to check the data because subclasses with other
+            # init-logic might try to init this class. We could skip these
+            # tests if we compared for self.__class__ but that has other
+            # drawbacks.
+
+            # Comparing if there is an explicit and an implicit unit parameter.
+            # If that is the case use the explicit one and issue a warning
+            # that there might be a conflict. In case there is no explicit
+            # unit just overwrite the unit parameter with the NDData.unit
+            # and proceed as if that one was given as parameter. Same for the
+            # other parameters.
+            if unit is not None and data.unit is not None:
+                log.info("Overwriting NDData's current "
```

Comment:
If this is done, don't you need to update the data values also, since the units are changing?

Also, it seems like a message is unnecessary if both units are the same, so maybe check for `and unit is not data.unit`?